### ISSUE TITLE
logback-core dependency should have `provided` scope

### ISSUE THREAD

**Issue by yborovikov** [logstash/logstash-logback-encoder#288]
Given that `logback-core` is indeed provided at runtime and README clearly states that as a (rather obvious) requirement, it would be cleaner to classify the former as `provided`-scope dependency.

This would also allow various custom/in-house libraries that extend `logstash-logback-encoder` to avoid introducing `logback-core` as transitive compile-scope dependency into their customers' class path (currently one has to specifically exclude `logback-core` to achieve that).

**Comment by philsttr** [comment]
Hi @yborovikov 

There are two options here:

1. (current behavior) logback-core is a `compile` scoped dependency of logback-logstash-encoder  
   * Users with libraries/apps that have a `compile` scoped dependency on logstash-logback-encoder, but don't want logback-core on their compile-time classpath can exclude logback-core when depending on logstash-logback-encoder.
   * Users with libraries/apps that have a `runtime` scoped dependency on logback-logstash-encoder will transitively depend on logback-core as well.  Therefore logback-logstash-encoder will work without worrying about including additional dependencies.  (logback-core might be provided by your runtime container, but it is not necessarily provided in every runtime environment)
2. (proposed behavior) logback-core is a `provided` scoped dependency of logback-logstash-encoder
   * Users with libraries/apps that have a `compile` scoped dependency on logstash-logback-encoder, but don't want logback-core on their compile-time classpath don't have to do anything.
   * Users with libraries/apps that have a `runtime` scoped dependency on logback-logstash-encoder will also need to include a `runtime` scoped dependency on logback-core in order for logback-logstash-encoder to work at runtime.


Given these two options, I prefer to keep the current behavior (1), since it favors libraries/apps that have a `runtime` scoped dependency on logstash-logback-encoder  (i.e. it makes their lives easier, since they don't have to remember to include an additional runtime dependency on logback-core).  There are more people that just use logstash-logback-encoder at runtime, than those that have libraries/apps that depend on logstash-logback-encoder for compile time extensions.  And remember that logback-core is not something that is provided by most containers by default.   I'd prefer for logback-logstash-encoder to work by just including logstash-logback-encoder as a dependency for these people.

In other words, I definitely recognize that there are people with compile-time dependencies on logstash-logback-encoder that do not also want logback-core included on their compile-time classpath.  It's just that this is a more complex use case than the basic use case, and I'd prefer the dependency scopes to favor the basic use case.

**Comment by yborovikov** [comment]
Hi @philsttr, thank you for your detailed explanation!

Striking a balance between convenience and correctness/flexibility is often a tricky challenge, indeed. In this particular case, however, it might be quite simple.

The reason is that for the majority of `logstash-logback-encoder` consumers (using it for logging) the consumers would _still_ have to explicitly declare either `logback-classic` or `logback-access` (or both) as their `runtime` scoped dependencies - simply relying on the transitive `logback-core` dependency of `logstash-logback-encoder` would not be enough in most cases I can think of (is there a popular way to use pure `logback-core` directly?).

Given that (and the possibility of `logback-core` version mismatch between customer customer-declared `logback-classic` and the one potentially coming from `logstash-logback-encoder`), I would argue that categorizing `logback-classic` as `provided` scope (in this library) would not harm the convenience, at all, in the typical basic use case.

**Comment by philsttr** [comment]
Hi @yborovikov,

You have convinced me.  I had overlooked that logback-classic/logback-access are both optional dependencies, and users have to add those dependencies anyway.

I will include this in the next release.

**Comment by yborovikov** [comment]
Thank you!
